lib/gameboard/board.rb

```diff
@@ -0,0 +1,495 @@
+    #                 2 | X | Y | Z |
+    #                   -------------
+    #                   - 0 - 1 - 2 -
+    #
+    #
+    def initialize(height:false, width: false, cells: false, preset: false)
+      raise ArgumentError unless ((height.is_a?(Integer) && width.is_a?(Integer)) || !!preset)
+      @height = height
+      @width = width
+      @cells = cells || nil
+      preset ? load_game(preset) : new_board
+    end
+
+    # Public: An enumerator that passes the individual cells.
+    #
+    # block - Individual cells have a :coord and a :value method where :coord
+    #         is a member of the Coordinate class
+    #
+    # Example
+    #
+    #   board.each { |cell| p cell }
+    #     #=> #<Gameboard::Cell:0x005637e5004790
+    #            @value = "O",
+    #            @coord = #<Gameboard::Coordinate:0x005637e5004a88
+    #                        @position = [0, 0], @x = 0, @y = 0>>
+    #
+    #     #=> #<Gameboard::Cell...
+    #
+    def each(&block)
+      return enum_for(__method__) if block.nil?
+      board.each(&block)
+      board
+    end
+
+    # Public: An enumerator that yields each cells value.
+    # Example
+    #
+    #   board.each_value { |cell| p cell }
+    #     #=> "O"
+    #     #=> "X"
+    #     #=> "X"...
+    #
+    def each_value(&block)
+      return enum_for(__method__) if block.nil?
+      each{|cell| yield(cell.value)}
+    end
+
+    # Public: An enumerator that yields each cells coordinates.
+    # Example
+    #
+    #   board.each { |cell| p cell }
+    #     #=> [0, 0]
+    #     #=> [1, 0]
+    #     #=> [2, 0]
+    #     #=> [0, 1]
+    #     #=> [1, 1]...
+    #
+    def each_coordinate(&block)
+      return enum_for(__method__) if block.nil?
+      each { |cell| yield(cell.coord.position) }
+    end
+
+    # Public: Return the value of the next available point given a coordinate
+    #         and a slope.
+    #
+    # point  - The starting coordinate pair.
+    # slope  - An [x, y] slope (e.g. [1,1] means over one row, up one column).
+    #          It must be an integer.
+    # coords - An optional boolean to specify returning a coordinate pair
+    #          instead of a value.
+    #
+    # Example
+    #   given_board =   -------------
+    #                 0 | 1 | 2 | 3 |
+    #                   -------------
+    #                 1 | A | B | C |
+    #                   -------------
+    #                 2 | X | Y | Z |
+    #                   -------------
+    #                   - 0 - 1 - 2 -
+    #
+    #   delta([1,1], [-1, 1])
+    #     #=> 1 ([0,2] if coord argument given)
+    #
+    #   delta([0,0], [2, 1])
+    #     #=> C ([2,1] if coord argument given)
+    #
+    def delta(point, slope, coords = false)
+      delta =  point.zip(slope).map {|point| point.reduce(:+) }
+      piece = board.find { |cell| cell.coord.position == delta }
+      raise "Off Grid" unless !!piece
+      coords ? piece.coord.position : piece.value
```

Comment:
Same point as above regarding position array vs Coordinate object.

I like the flexibility of returning coordinates versus values, but there seems to be no way to return the actual cell. I think returning the Cell object itself should be the default behavior.

Users can always deconstruct the cell later without much difficulty, a simple `map` would do the trick.

I could imagine many cases where the user would need to make some decision based on the value and would then need the coordinate to enact that decision upon the board.